### DISCUSSION TITLE
error loading MSBuildWorkspace for projects containing .xaml files

### DISCUSSION THREAD

**Discussion by ltcmelo** [dotnet/roslyn#45642] [General]
**Version Used**: 3.6.0

I'm unable to correctly `OpenSolutionAsync` through `MSBuildWorkspace` when a project contains **.xaml** files. I get the following load failure:

     Msbuild failed when processing the file 'C:\...\Name.csproj' with message: C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets: (268, 9): Unknown build error, 'Object reference not set to an instance of an object.'

Which consequently leads to compilation errors like:

    C:\..\MainWindow.xaml.cs(25,13): error CS0103: The name 'InitializeComponent' does not exist in the current context

I've found [this](https://github.com/dotnet/roslyn/issues/2779) and [this](https://github.com/dotnet/roslyn/pull/2982) old issues (from 2015) that look the same as the one I'm facing. These, along with additional resources I've found online, suggest the setting the MSBuild properties `DesignTimeBuild` and `BuildingInsideVisualStudio` to `true` (in the dictionary that's passed to `MSBuildWorkspace.Create` would fix the issue. However, it seems that those properties are already set by default in the newer `MSBuildWorkspace` versions (I'm using **3.6.0**), so they didn't have any effect.

Other things I've tried as well:

- setting `CheckForSystemRuntimeDependency` to `true`;
- setting `ws.LoadMetadataForReferencedProjects`  to `true` (where `ws` is the workspace created).

None of the above helped.

This is a snippet of how the `.csproj` includes the `.xaml` content.

```
<ItemGroup>
    <Compile Include="MainWindow.xaml.cs" />
    ...
</ItemGroup>
<ItemGroup>
    <Page Include="MainWindow.xaml">
        <Generator>MSBuild:Compile</Generator>
        <SubType>Designer</SubType>
    </Page>
</ItemGroup>
```

An invocation of msbuild.exe from the command line successfully builds the project.

What's the recommendation to deal with this kind of artefacts in a project? (I'd accept a "unorthodox" way of compiling the hidden `.g.i.cs` files, as long as it's **robust/reliable**).

**Comment by softwaretirol** [comment]
I would be intrested too,.. i could successfully emit an assembly out of a project which is .NET Core based and contains some XAML Files if I am passing "BuildingInsideVisualStudio". 

In this case, the ".g.cs" files are getting generated, but i am missing the resource files with the compiled (B)XAML content? In case I am removing the "BuildingInsideVisualStudio", i got the same exception as @ltcmelo  

```csharp
using var workspace = MSBuildWorkspace.Create(new Dictionary<string, string>()
{
    { "BuildingInsideVisualStudio", "true" }
});
workspace.WorkspaceFailed += (o, e) => Console.WriteLine(e.Diagnostic.Message);
workspace.LoadMetadataForReferencedProjects = true;
var project = await workspace.OpenProjectAsync(sourceProject, new ConsoleLogger());
var compilation = await project.GetCompilationAsync();
var res = compilation.Emit("Test.exe");
```

**Comment by bert2** [comment]
I'm using 3.7.0 and I've hit the same issue. I got passed that NRE, but ran into another error. This property helps ([source](https://github.com/OmniSharp/omnisharp-roslyn/pull/1157)):

```csharp
var workspace = MSBuildWorkspace.Create(new Dictionary<string, string> {
    ["AlwaysCompileMarkupFilesInSeparateDomain"] = "false"
});
```

The new error is: 

> [Failure] Msbuild failed when processing the file 'D:\DEV\microscope\src\VSExtension\VSExtension.csproj' with message: C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets: (268, 9): Unknown build error, 'Could not load file or assembly 'Microsoft.CodeAnalysis.Workspaces, Version=3.7.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)' .

[Running the build task "manually"](https://stackoverflow.com/questions/24151954/compile-xaml-using-roslyn-to-get-g-cs-file) might help.

Fortunately for me, I can just move my XAML files into a different project, because I don't actually have to compile them with the `MSBuildWorkspace`. Good luck to you as this looks like a very messy situation...
